docs/2019-03-28-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,240 @@
+Proving All-Path Reachability Claims
+====================================
+
+This document details All-Path Reachability without solving the
+most-general-unifier (MGU) problem.
+MGU will be detailed in a separate document.
+
+Background
+----------
+
+Recommended reading: [proving One-Path Reachability
+claims](2018-11-08-One-Path-Reachability-Proofs.md).
+
+
+Definitions
+-----------
+
+### Weak always
+
+Given a formula `φ`, let `[w]φ` denote the formula “weak always” `φ`,
+defined by:
+
+```
+[w]φ := νX.φ ∨ (○X ∧ •⊤)
+```
+
+Problem Description
+-------------------
+
+Given a set of claims of the form `∀x.φ → [w]ψ`, we are trying to prove
+all of them together.
+
+
+Algorithm
+---------
+
+(explained in the next section)
+
+__Input:__ claims `∀x₁.φ₁ → [w]ψ₁`, `∀x₂.φ₂ → [w]ψ₂`, …, `∀xₙ.φₙ → [w]ψₙ`
+
+__Output:__ Provable or Unprovable
+
+* For each claim `∀x.φ → [w]ψ`
+    * While φ is not ⊥ :
+        * Let `φ’ := φ ∧ ¬∃x.⌈φ ∧ ψ⌉` 
+        * If `φ’ = ⊥` Then exit while
+            (continue to the next claim)
+        * If not the first while iteration for this claim:
+            * Apply all claims on `φ’` as we would apply `∀x.φ → <w>ψ` claims
+                in the one-path algorithm (see above)
+            * Output consists of a disjunction of rewrite results
+                and a remainder (the un-rewritten part of `φ’`).
+            * Set `φ’` to be this remainder
+        * For all axioms `∀y.α → •β`:
+            * Let `γ(α) := ∃y.⌈φ’ ∧ α⌉` be the MGU of `φ’` and `α` 
+            * Let `β ∧ γ(α)` be the rewrite result for this axiom.
+        * If `φ’ ∧ ¬γ(α₁) ∧ …  ¬γ(αₙ) ≠ ⊥`, 
+            where `αᵢ`s are the lhs patterns of our rewrite axioms above:
+            * Return `Unprovable`
+        * Set `φ` to be the disjunction of all rewrite results above
+            (including those from claims, if any)
+* Return `Provable`
+
+
+Explanation
+-----------
+
+Say we want to prove `φ → [w]ψ`.
+
+Unrolling `[w]ψ` we obtain `φ → ψ ∨ ○[w]ψ ∧ •⊤`.
+
+Moving `ψ` to the left of the implication, we get the equivalent
+
+```
+φ ∧ ¬ψ →  ○[w]ψ ∧ •⊤
+```
+
+Let `φ'` be `φ ∧ ¬ψ`. This step eliminates the cases in which `ψ` holds now.
+
+If `φ'` is `⊥`, then the implication holds and we are done.
+
+### Simplifying `φ ∧ ¬∃x.ψ` where x does not appear in `φ`
+
+```
+φ ∧ ¬∃x.ψ                                           is equivalent with
+φ ∧ ¬∃x.ψ /\ (∃x.⌈φ∧ψ⌉ ∨ ¬∃x.⌈φ∧ψ⌉)                 is equivalent with
+(φ ∧ ¬∃x.ψ /\ ∃x.⌈φ∧ψ⌉) ∨ (φ ∧ ¬∃x.ψ ∧ ¬∃x.⌈φ∧ψ⌉)   is equivalent with
+(φ ∧ ∃x.ψ ∧ ¬∃x.ψ) ∨ (φ ∧ ¬∃x.ψ ∧ ¬∃x.⌈φ∧ψ⌉)        is equivalent with
```

Comment:
assuming \phi is partial functional?

Reply:
Yes, I should probably say somewhere that we are under the same assumption as for the one-path reachability design doc